python/pyarrow/tests/test_pandas.py

```diff
@@ -149,6 +149,12 @@ def test_non_string_columns(self):
         table = pa.Table.from_pandas(df)
         assert table.field(0).name == '0'
 
+    def test_non_string_columns_with_index(self):
+        df = pd.DataFrame({0: [1, 2, 3], 1: [4, 5, 6]})
+        df = df.set_index(0)
+        table = pa.Table.from_pandas(df)
+        assert table.field(0).name == '1'
```

Comment:
I suppose, assuming round-tripping the type of the index names works, we should also `_check_pandas_roundtrip(df, preserve_index=True)` here

Reply:
Good point! I tested that `_check_pandas_roundtrip` will fail